lib/path-design.md

```diff
@@ -0,0 +1,373 @@
+- Joining a single path is that path itself, but normalised:
+  `relative.join [ p ] == normalise p`
+
+### `relative.normalise`
+
+```haskell
+relative.normalise :: String -> String
+```
+
+Normalise relative paths.
+
+- Limit repeating `/` to a single one (does not change a [POSIX Pathname](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_271))
+- Remove redundant `.` components (does not change the result of [POSIX Pathname Resolution](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_13))
+- Error on empty strings (not allowed as a [POSIX Filename](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_170))
+- Remove trailing `/` and `/.` (See [justification](#trailing-slashes))
+- Error on `..` path components (See [justification](#parents))
+- Remove leading `./` (See [justification](#leading-dots))
+
+Examples:
+```
+# limit repeating `/` to a single one
+relative.normalise "foo//bar" == "foo/bar"
+
+# remove redundant `.` components
+relative.normalise "foo/./bar" == "foo/bar"
+
+# remove leading `./`
+# TODO: bikeshedding. there is a use to this.
+relative.normalise "./foo/bar" == "foo/bar"
+
+# remove trailing `/`
+relative.normalise "foo/bar/" == "foo/bar"
+
+# remove trailing `/.`
+relative.normalise "foo/bar/." == "foo/bar"
+
+# error on `..` path components
+relative.normalise "foo/../bar" == <error>
+
+# error on empty string
+relative.normalise "" == <error>
+
+# error on absolute path
+relative.normalise "/foo" == <error>
+```
+
+Laws:
+- Idempotency:
+
+  `relative.normalise (relative.normalise p) == relative.normalise p`
+
+- Doesn't change the file system object pointed to:
+
+  `$(stat ${p}) == $(stat ${relative.normalise p})`
+
+  TODO: Is this the same as the law below?
+
+- Uniqueness: If the normalisation of two paths is different then they point to different paths:
+
+  `normalise p != normalise q => $(stat ${p}) != $(stat ${q})`
+
+  Note: This law only holds if trailing slashes are not persisted, see the [trailing slashes decision](#trailing-slashes)
+
+### `difference`
+
+```haskell
+difference :: AttrsOf Path -> { commonPrefix :: Path; suffix :: AttrsOf String; }
+```
+
+Take the difference between multiple paths, returning the common prefix between them and the respective suffices.
+
+Examples:
+```nix
+difference { path = /foo/bar } = { commonPrefix = /foo/bar; suffix = { path = "."; }; }
+difference { left = /foo/bar; right = /foo/baz; } = { commonPrefix = /foo; suffix = { left = "bar"; right = "baz"; }; }
+difference { left = /foo; right = /foo/bar; } = { commonPrefix = /foo; suffix = { left = "."; right = "bar"; }; }
+difference { left = /.; right = /foo; } = { commonPrefix = /.; suffix = { left = "."; right = "bar"; }; }
+difference { left = /foo; right = /foo; } = { commonPrefix = /foo; suffix = { left = "."; right = "."; }; }
+
+# Requires at least one path
+difference {} = <error>
```

Comment:
This function reminds me more of the _greatest common divisor_ than _difference_.

The word _difference_ applied to paths makes me think of set difference.

**tangent**

..., which something I've considered implementing but purposefully haven't
 - https://github.com/NixOS/nixpkgs/pull/112083
 
 (noting that trees are quite different from sets of paths)

Set difference is not very different from `sources.filter` and might suggest a usage that may not perform well.